src/main/java/frc/robot/subsystems/swerve/SwerveSubsystem.java

```diff
@@ -305,71 +304,72 @@ public void periodic() {
       // Apply update
       estimator.updateWithTime(sampleTimestamps[deltaIndex], rawGyroRotation, modulePositions);
     }
-    Logger.recordOutput("Swerve Pose", pose);
 
-    for (var camera : cameras) {
-      PhotonPipelineResult result =
-          new PhotonPipelineResult(camera.inputs.latency, camera.inputs.targets);
-      result.setTimestampSeconds(camera.inputs.timestamp);
-      try {
-        var estPose =
+    if (Robot.isReal()) {
+      List<Pose3d> visionPoses = new ArrayList<>();
+      for (var camera : cameras) {
+        PhotonPipelineResult result =
+            new PhotonPipelineResult(camera.inputs.latency, camera.inputs.targets);
+        result.setTimestampSeconds(camera.inputs.timestamp);
+        try {
+          var estPose =
+              VisionHelper.update(
+                      result,
+                      camera.constants.cameraMatrix(),
+                      camera.constants.distCoeffs(),
+                      PoseStrategy.MULTI_TAG_PNP_ON_COPROCESSOR,
+                      fieldTags,
+                      camera.constants.robotToCamera())
+                  .get();
+          var visionPose = estPose.estimatedPose;
+          visionPoses.add(visionPose);
+          Logger.recordOutput(
+              "Vision/Vision Pose From " + camera.constants.cameraName(), visionPose);
+          estimator.addVisionMeasurement(
+              visionPose.toPose2d(),
+              camera.inputs.timestamp,
+              VisionHelper.findVisionMeasurements(estPose));
```

Comment:
this could be named better, should include StdDeviations or something along those lines